isaac/controllers.md

```diff
@@ -0,0 +1,255 @@
+If there are any properties or invariants the object is supposed to maintain, there are the controllers that may have "messed up" the object. 
+In the simplest case, we might imagine tagging each object with a set of _affecting controllers_.
+In general, if an application (or a user) trusts all the affecting controllers in an object's tag, it can rely on that object.
+A simple but effective set of rules would be:
+- When an object is created, its affecting controllers set is whatever controller is authoritative for the object.
+- When an object transfers to a new controller, we add the new controller to its affecting controllers.
+- If an object updates its state based on the state of other objects (e.g. receiving a bank transfer from another account), we add the affecting controllers from all dependency objects to this object's affecting controllers. 
+This solution is very similar to [taint tracking](http://www.tcse.cn/~wsdou/papers/2022-dsn-dista.pdf) or in Programming Language & Distributed Systems research: each object is "tainted" by its affecting controllers, and contributes that "taint" to everything that depends on it. 
+
+The tricky part is in _removing_ controllers from objects' tags.
+A common case in blockchain systems is to move an object from some chain (let's call it _base chain_) to some other chain (let's call it _L2_), do some stuff, and then move the object back to _base chain_. 
+We then want _base chain_ to somehow "review" the changes made on _L2_, and somehow be able to treat the object like it was never on _L2_ at all.
+We need to remove _L2_ from the object's tag.
+In [Information Flow Control](https://en.wikipedia.org/wiki/Information_flow_(information_theory)) research, this is similar to _endorsement_, [a fairly well-explored problem](https://doi.org/10.1145/3133956.3134054).
+
+
+
+<figure>
+    <img src="media/wrapping-unwrapping.png" alt="wrapping and unwrapping a token as it's transferred across chains" />
+  <figcaption>with [IBC](https://github.com/cosmos/ibc/blob/8d9d3b6fe7309b034df8457760b3bbd11d24b8e1/archive/papers/2020-05/build/paper.pdf#L2) and [ICS20](https://github.com/cosmos/ibc/blob/main/spec/app/ics-020-fungible-token-transfer/README.md), users can transfer tokens between chains, "destroying" the original, and creating "wrapped" tokens, which are "unwrapped" when they are transferred back. 
+</figcaption>
+</figure>
+TODO: better IFC-style wrapping/unwrapping diagram
+
+The blockchain setting, however, introduces a unique twist: instead of relying on _trust_ as the main basis for endorsement, modern blockchains can have the power to _review_ and _verify_ histories of other chains, using [recursive ZKPs](https://eprint.iacr.org/2021/370), [optimistic proofs](https://www.coindesk.com/tech/2024/03/19/optimism-finally-starts-testing-fault-proofs-at-heart-of-design-and-of-criticism/), or simply fully replicating the other chain's state machine.
```

Comment:
It might be helpful for the reader to name some examples, especially the recursive ZKP interoperability techniques; 
- [Agg Layer](https://mirror.xyz/0xfa892B19c72c2D2C6B10dFce8Ff8E7a955b58A61/TXMyZhhRFa-bjr7YHwmJpKBwt2-_ysirbh_VpNy3qZY)
- ZK IBC 

or any others you know of. The reason it's important is that the [industry is starting to realize](https://www.youtube.com/watch?v=CF-F-jbA630&pp=ygUcYnJpZGdlcyBhcmUgYSBzY2FtIHNodW1vIGNodQ%3D%3D) that [ZK will be crucial](https://www.youtube.com/watch?v=b_ElApwRvkY&pp=ygUSYmVsbCBjdXJ2ZSB1bWEgcm95) for interoperability going forward. As an example, my own [naïveté](https://research.anoma.net/t/controllers-from-first-principles/66/2?u=apriori)